src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,245 @@
+/*
+ *
+ * Copyright 2016 CUBRID Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * memory_monitor_sr.hpp - memory monitoring module header
+ */
+
+#ifndef _MEMORY_MONITOR_SR_HPP_
+#define _MEMORY_MONITOR_SR_HPP_
+
+#if !defined (SERVER_MODE)
+#error Belongs to server module
+#endif /* SERVER_MODE */
+
+#include <cstring>
+#include <type_traits>
+#include <cstdint>
+#include <atomic>
+#include <vector>
+#include <cassert>
+
+#include "perf_def.hpp"
+#include "thread_compat.hpp"
+#include "memory_monitor_common.h"
+
+#define MMM_PARSE_MASK 0x0000FFFF
+#define MMM_MAKE_MODULE_INIT_STAT_ID(module_idx) ((module_idx) << 16)
+#define MMM_MODULE_IDX_FROM_STAT_ID(stat) ((stat) >> 16)
+#define MMM_COMP_INFO_IDX_FROM_STAT_ID(stat) ((stat) & MMM_PARSE_MASK)
+
+typedef enum
+{
+  HEAP = MMM_MAKE_MODULE_INIT_STAT_ID (MMM_HEAP_MODULE),
+  HEAP_ATTRINFO,
+  MMM_STAT_LAST = MMM_MAKE_MODULE_INIT_STAT_ID (MMM_MODULE_LAST)
+} MMM_STAT_ID;
+
+int memmon_add_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+int memmon_sub_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t size);
+int memmon_move_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID src, MMM_STAT_ID dest, uint64_t size);
+int memmon_resize_stat (THREAD_ENTRY *thread_p, MMM_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+
+namespace cubperf
+{
+  class memory_monitoring_manager;
+  class mmm_aggregater;
+
+  typedef struct mmm_comp_info
+  {
+    MMM_STAT_ID idx;
+    std::string comp_name;
+    std::string subcomp_name;
+  } MMM_COMP_INFO;
+
+  typedef struct mmm_mem_stat
+  {
+    std::atomic<uint64_t> init_stat;
+    std::atomic<uint64_t> cur_stat;
+    std::atomic<uint64_t> peak_stat;
+    std::atomic<uint32_t> expand_count;
+  } MMM_MEM_STAT;
+
+  MMM_COMP_INFO heap_comp_info[] =
+  {
+    {HEAP, "", ""},
+    {HEAP_ATTRINFO, "ATTRINFO CACHE", ""}
+  };
+
+  class mmm_subcomponent
+  {
+    public:
+      mmm_subcomponent (const char *name)
+      {
+	m_subcomp_name = new char[strlen (name) + 1];
+	strcpy (m_subcomp_name, name);
+      }
+
+      /* variable */
+      char *m_subcomp_name;
+      std::atomic<uint64_t> m_cur_stat;
```

Comment:
The member variables functions have to be `private` unless it has to be open. They will be used outside? Please consider it through the whole code.

Reply:
I modify it in 334d324.